### PR TITLE
Validate before creating a new deployment

### PR DESCRIPTION
**Pull request by danafallon**
Also created a new struct called `BaseDeployment` which doesn't include the fields that are only set by our api (uuid and status) - this allows us to construct an api model for a deployment that hasn't been created yet. Now the client methods that can work on an unsaved deployment take a `BaseDeployment` instead of a full `Deployment`.

Will make the same change for destinations and ssh tunnels after this is merged.

### CODE REVIEW

File: internal/provider/deployment_resource.go [hunk 1]
```diff
@@ -189,22 +189,32 @@ func (r *DeploymentResource) Create(ctx context.Context, req resource.CreateRequ
...
 	}
 
-	// Translate the Terraform plan into an API model and then fill in computed fields
-	// from the API response of the newly created deployment
-	model := models.DeploymentResourceToAPIModel(data)
-	model.UUID = deployment.UUID
-	model.Status = deployment.Status
+	// Fill in computed fields from the API response of the newly created deployment
+	fullDeployment := models.BaseDeploymentAPIModelToDeploymentAPIModel(deployment, createdDeployment.UUID, createdDeployment.Status)
+	fullDeployment.Status = createdDeployment.Status
```
Comment (nathan-artie) [inline internal/provider/deployment_resource.go:214]: Does status need to be set since it's passed into `BaseDeploymentAPIModelToDeploymentAPIModel`?

Comment (danafallon) [reply internal/provider/deployment_resource.go:214]: Oh yeah I meant to remove this line, good catch

File: internal/provider/deployment_resource.go [hunk 2]
```diff
@@ -189,22 +189,31 @@ func (r *DeploymentResource) Create(ctx context.Context, req resource.CreateRequ
...
+
+	if err := r.client.Deployments().ValidateDestination(ctx, deployment); err != nil {
+		resp.Diagnostics.AddError("Unable to Create Deployment", err.Error())
+		return
+	}
+
 	// Our API's create endpoint only accepts the source type, so we need to send two requests:
 	// one to create the bare-bones deployment, then one to update it with the rest of the data
-	deployment, err := r.client.Deployments().Create(ctx, data.Source.Type.ValueString())
+	createdDeployment, err := r.client.Deployments().Create(ctx, data.Source.Type.ValueString())
```
Comment (nathan-artie) [inline internal/provider/deployment_resource.go:206]: Could you use `deployment.Source.Type` here?

File: internal/provider/deployment_resource.go [hunk 3]
```diff
@@ -189,22 +189,31 @@ func (r *DeploymentResource) Create(ctx context.Context, req resource.CreateRequ
...
 		return
 	}
 
-	// Translate the Terraform plan into an API model and then fill in computed fields
-	// from the API response of the newly created deployment
-	model := models.DeploymentResourceToAPIModel(data)
-	model.UUID = deployment.UUID
-	model.Status = deployment.Status
+	// Fill in computed fields from the API response of the newly created deployment
+	fullDeployment := deployment.ToFullDeployment(createdDeployment.UUID, createdDeployment.Status)
```
Comment (nathan-artie) [inline internal/provider/deployment_resource.go:213]: Could you just use `createdDeployment` from here out?

File: internal/provider/deployment_resource.go [hunk 4]
```diff
@@ -242,25 +251,26 @@ func (r *DeploymentResource) Update(ctx context.Context, req resource.UpdateRequ
...
 
-	if err := r.client.Deployments().ValidateDestination(ctx, deployment); err != nil {
+	if err := r.client.Deployments().ValidateDestination(ctx, baseDeployment); err != nil {
 		resp.Diagnostics.AddError("Unable to Update Deployment", err.Error())
 		return
 	}
 
-	deployment, err := r.client.Deployments().Update(ctx, deployment)
+	fullDeployment := data.ToAPIModel()
+	updatedDeployment, err := r.client.Deployments().Update(ctx, fullDeployment)
```
Comment (nathan-artie) [inline internal/provider/deployment_resource.go:266]: Could inline.

**Review by nathan-artie — APPROVED** [review]

**Pull request merged**